### DISCUSSION TITLE
Why are imported modules on serverless functions undefined?

### DISCUSSION THREAD

**Discussion by qodesmith** [vercel/community#893] [Help] [answered]
_TL;DR - why do imports return `undefined`? Especially for native Node modules like `path`?_

I have a very small test application, built with Vite, that has a single endpoint in the `/api` directory. I created this just to play around with Vite + Vercel.

I have only 2 module imports for my endpoint - `path` and `fs-extra`. Both are returning as `undefined`. I was getting `cannot read join of undefined` errors with the `path` module, so I wrapped everything in a `try/catch` just to see if the endpoint responds. It does. See my code below.

```typescript
import type {VercelRequest, VercelResponse} from '@vercel/node'
import path from 'node:path'
import fs from 'fs-extra'

export default function handler(req: VercelRequest, res: VercelResponse) {
  // Both of these log 'undefined' on my Vercel dashboard for function logs.
  console.log('path module:', path)
  console.log('fs module:', fs)

  try {
    // https://vercel.com/guides/how-can-i-use-files-in-serverless-functions
    const pathFromProjectRootToFile = '/api/usersData.json'
    const usersDataFilePath = path.join( // THIS IS WHERE THE ERROR HAPPENS 🙃
      process.cwd(),
      pathFromProjectRootToFile
    )
    const usersData = fs.readJSONSync(usersDataFilePath)
    res.json({users: usersData})
  } catch (e) {
    res.json({error: errorToObject(e as Error)})
  }
}

function errorToObject(err?: Error): Record<string, string> | null {
  if (!err || !(err instanceof Error)) return null

  const obj: Record<string, string> = {}

  Object.getOwnPropertyNames(err).forEach(prop => {
    const value = err[prop as keyof Error]

    if (typeof value !== 'string') return

    obj[prop] = value
  })

  return obj
}
```

As an aside, instead of `node:path` I also tried just `path`, but same thing - `undefined`. And I do have `fs-extra` as a dependency in my `package.json`.

**Comment by amyegan** [comment, marked as answer]
Hi @qodesmith! I have a few questions to help guide us through debugging this problem.
- Is `fs-extra` installed under devDependencies or  normal Dependencies?
- Does it also fail to find the basic fs import, like `import fs from "fs"`? 
- Are you using regular [Serverless Functions](https://vercel.com/docs/concepts/functions/serverless-functions), or are you using [Edge Functions](https://vercel.com/docs/concepts/functions/edge-functions)? _(Edge Runtime does not have access to native Node.js APIs like process, path, or fs.)_
- Are you able to share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) to let me take a closer look into what is going wrong?

**Reply by qodesmith** [reply]
Hi @amyegan, I appreciate your efforts with answering questions in these community discussions. I have published a [repo on Github](https://github.com/qodesmith/vite-vercel-example) that reproducing the problem. The app deployed on Vercel can be seen [here](https://vite-vercel-example-sqn5.vercel.app/).

Let me answer your questions:
* > Is fs-extra installed under devDependencies or normal Dependencies?
  * Dependencies.
* > Does it also fail to find the basic fs import, like import fs from "fs"?
  * Yes. I just recently added the `fs` module and logged its value in the serverless function. I've attached a screenshot below. All 3 modules are undefined - `path`, `fs`, and `fs-extra`.
  * [Here's a link](https://github.com/qodesmith/vite-vercel-example/blob/main/api/users.ts#L8-L10) to the code where I'm importing these modules and `console.log`'ing their values.
  * <img width="1230" alt="Screen Shot 2022-09-27 at 7 46 28 AM" src="https://user-images.githubusercontent.com/11034792/192518862-1eea3bee-c2f9-417f-88b4-b02efc0d5f55.png">
* > Are you using regular [Serverless Functions](https://vercel.com/docs/concepts/functions/serverless-functions), or are you using [Edge Functions](https://vercel.com/docs/concepts/functions/edge-functions)? (Edge Runtime does not have access to native Node.js APIs like process, path, or fs.)
  * I _thought_ I was using regular serverless functions. This is a Vite-created app. The Vercel docs said that if you want to add serverless functions as a back end just add an `/api` directory at the root of the project. When deployed on Vercel, it picks it up and deploys the functions. Is there a way to verify the functions are regular serverless vs. edge functions?
* >  Are you able to share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) to let me take a closer look into what is going wrong?
  * Yes. The repo can be found [here](https://github.com/qodesmith/vite-vercel-example) and the Vercel deployed link to the app is [here](https://vite-vercel-example-sqn5.vercel.app/).

**Reply by amyegan** [reply]
Thank you for sharing this! I believe the source of the problem is the file location. Storing the JSON data file in `/api` isn't working because that is a special directory treated differently from the others. If you move the data into another directory and import from there, it will work. 

PR with these adjustments: https://github.com/qodesmith/vite-vercel-example/pull/1

**Reply by qodesmith** [reply]
@amyegan Thanks for the PR. I just merged but all 3 modules - `path`, `fs`, and `fs-extra` - are still coming in as undefined:
<img width="1230" alt="Screen Shot 2022-09-28 at 10 32 42 AM" src="https://user-images.githubusercontent.com/11034792/192807038-20bffc13-6aa3-4d50-b7b4-480f124999d1.png">

You mentioned earlier that edge functions don't have access to native modules. How can we verify that these functions are serverless and not edge? It's still a mystery why importing any module results in `undefined`.

**Reply by amyegan** [reply]
🤦 I'm an idiot. Sorry for my mistake! I'll blame Hurricane Ian for distracting me here in Florida 😅  

The `import` needed to be changed to `require` to work with Vite. You can see it working at this temporary deployment: https://qodesmith-vite-example.vercel.app/

New PR showing what I changed to make it actually work when deployed on Vercel: https://github.com/qodesmith/vite-vercel-example/pull/2

**Reply by amyegan** [reply]
Regarding [Edge Functions](https://vercel.com/docs/concepts/functions/edge-functions): You would need to [add a bit of configuration](https://vercel.com/docs/concepts/functions/edge-functions#creating-edge-functions) to be using them instead of Serverless Functions, so you're all set on that.

**Reply by qodesmith** [reply]
@amyegan Wow, I would have _never_ though of using require instead! Thanks so much, this was really hanging me up. Is there any official documentation around this? Out of everything that I was searching online, I haven't run across anything mentioning import vs require. Again, thank you so much!

**Reply by amyegan** [reply]
I don't think we have any documentation on that yet, so thank you for pointing this out! I'm adding it to the documentation suggestions list

**Comment by sunmer** [comment]
Can we get some official docs on how to do imports in a Vercel serverless and Vite environment? @amyegan
